curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -0,0 +1,1092 @@
+            /*
+                 -----               ReadData
+                    ------           DataCache
+            */
+            if (chunkPos + readLen <= dcChunkPos + dcLen) {
+                memcpy(dataBuf + request.len, cacheData,
+                       chunkPos + readLen - dcChunkPos);
+                readLen = 0;
+                break;
+                /*
+                     -----------         ReadData
+                        ------           DataCache
+                */
+            } else {
+                memcpy(dataBuf + request.len, cacheData, dcLen);
+                readLen = chunkPos + readLen - (dcChunkPos + dcLen);
+                dataBufOffset = dcChunkPos + dcLen - chunkPos + dataBufOffset;
+                chunkPos = dcChunkPos + dcLen;
+            }
+        } else if ((chunkPos >= dcChunkPos) &&
+                   (chunkPos < dcChunkPos + dcLen)) {
+            char* cacheData = (*dcpIter)->GetData();
+            s3ClientAdaptor_->GetFsCacheManager()->Get(iter->second);
+            /*
+                     ----              ReadData
+                   ---------           DataCache
+            */
+            if (chunkPos + readLen <= dcChunkPos + dcLen) {
+                memcpy(dataBuf, cacheData + chunkPos - dcChunkPos, readLen);
+                readLen = 0;
+                break;
+                /*
+                         ----------              ReadData
+                       ---------                DataCache
+                */
+            } else {
+                memcpy(dataBuf, cacheData + chunkPos - dcChunkPos,
+                       dcChunkPos + dcLen - chunkPos);
+                readLen = chunkPos + readLen - dcChunkPos - dcLen;
+                dataBufOffset = dcChunkPos + dcLen - chunkPos + dataBufOffset;
+                chunkPos = dcChunkPos + dcLen;
+            }
+        } else {
+            continue;
+        }
+    }
+
+    if (readLen > 0) {
+        ReadRequest request;
+        request.index = index_;
+        request.len = readLen;
+        request.chunkPos = chunkPos;
+        request.bufOffset = dataBufOffset;
+        requests->emplace_back(request);
+    }
+    return;
+}
+
+DataCachePtr ChunkCacheManager::FindWriteableDataCache(
+    uint64_t chunkPos, uint64_t len,
+    std::vector<DataCachePtr>* mergeDataCacheVer) {
+    // todo performance optimization
+    WriteLockGuard writeLockGuard(rwLockWrite_);
+    auto iter = dataWCacheMap_.begin();
+    uint32_t i = 0;
+    for (; iter != dataWCacheMap_.end(); iter++) {
+        LOG(INFO) << "FindWriteableDataCache pos:" << chunkPos
+                  << ", len:" << len
+                  << ", datacache pos:" << iter->second->GetChunkPos()
+                  << ", datacache len:" << iter->second->GetLen();
+        if (((chunkPos + len) >= iter->second->GetChunkPos()) &&
+            (chunkPos <=
+             iter->second->GetChunkPos() + iter->second->GetLen())) {
```

Comment:
since the map's key is chunk position, maybe we can use std::map 's function like find or upper_bound to speed up search progress?

Reply:
fixed